tests/validators/string.test.ts

```diff
@@ -95,5 +95,159 @@ describe('StringValidator', () => {
 				);
 			});
 		});
+
+		describe('email', () => {
+			const emailPredicate = s.string.email;
+
+			test.each(['hi@hello.com', 'foo@bar.net', 'hello+world@example.com'])('GIVEN %s THEN returns given value', (input) => {
+				expect(emailPredicate.parse(input)).toBe(input);
+			});
+
+			test.each(['hi@hello', 'foo@bar'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+				expect(() => emailPredicate.parse(input)).toThrow(
+					new ConstraintError('s.string.email', 'Invalid email address', input, 'expected to be an email address')
+				);
+			});
+		});
+
+		describe('url', () => {
+			describe('Without any options', () => {
+				const urlPredicate = s.string.url();
+
+				test.each(['https://google.com', 'http://foo.bar'])('GIVEN %s THEN returns given value', (input) => {
+					expect(urlPredicate.parse(input)).toBe(input);
+				});
+
+				test.each(['google.com', 'foo.bar'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => urlPredicate.parse(input)).toThrow(new ConstraintError('s.string.url', 'Invalid URL', input, 'expected.url'));
+				});
+			});
+
+			describe('With protocol', () => {
+				const urlPredicateWithProtocol = s.string.url({ allowedProtocols: ['git:'] });
+
+				test.each(['git://foo.bar'])('GIVEN %s THEN returns given value', (input) => {
+					expect(urlPredicateWithProtocol.parse(input)).toBe(input);
+				});
+
+				test.each(['https://google.com', 'http://foo.bar'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => urlPredicateWithProtocol.parse(input)).toThrow(
+						new ConstraintError('s.string.url', 'Invalid URL protocol', input, 'expected to be one of: git:')
+					);
+				});
+			});
+
+			describe('With domain', () => {
+				const urlPredicateWithDomain = s.string.url({ allowedDomains: ['google.com'] });
+
+				test.each(['https://google.com', 'http://google.com'])('GIVEN %s THEN returns given value', (input) => {
+					expect(urlPredicateWithDomain.parse(input)).toBe(input);
+				});
+
+				test.each(['https://foo.bar', 'http://foo.bar'])('GIVEN %s THEN throws a ConstraintError', (input) => {
+					expect(() => urlPredicateWithDomain.parse(input)).toThrow(
+						new ConstraintError('s.string.url', 'Invalid URL domain', input, 'expected to be one of: google.com')
+					);
+				});
+			});
+		});
+
+		describe('uuid', () => {
+			describe('default', () => {
+				const uuidPredicate = s.string.uuid();
+
+				test.each(['450d6a23-9e6f-45d9-9d5a-fd4f6e014f16', '00000000-0000-0000-0000-000000000000'])(
```

Comment:
UUIDs should move to variables `uuid4`, `uuid6`, and `uuidNull`, all of which should defined within `describe('uuid', cb)`.